### PR TITLE
add support for tenancy and the developer persona

### PR DESCRIPTION
**Pull request by tchughesiv**
Screen Shot 2021-08-27 at 2 05 02 PM

### CODE REVIEW

File: src/components/instanceListPage.jsx [hunk 1]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
     }
   }
 
-  const fetchInstances = () => {
+  async function fetchInventoryItems() {
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:136]: @tchughesiv Minor: maybe consider to rename the function to something meaningful? For example, "fetchInventoryByNSAndRules"?

File: src/components/instanceListPage.jsx [hunk 2]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+    }
+    await fetch('/api/kubernetes/apis/authorization.k8s.io/v1/selfsubjectrulesreviews', requestOpts)
+      .then(status)
+      .then(json)
+      .then(parseTenantRulesReview)
+      .then(fetchInventoryNamespaces)
+      .then(fetchInventoriesByNamespace)
+      .then((inventories) => inventoryItems.push(...inventories))
+      .catch(function (error) {
+        console.error(error)
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:161]: @tchughesiv Should we show the error msg to users? If so, we can do
`setFetchInstancesFailed(true)`
`setStatusMsg(error)`

File: src/components/instanceListPage.jsx [hunk 3]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+      promises.push(
+        fetchTenant(tenantName).then((data) => {
+          return data.spec.inventoryNamespace
+        })
+      )
+    })
+    await Promise.all(promises)
+      .then((namespaces) => (inventoryNamespaces = namespaces))
+      .catch(function (error) {
+        console.error(error)
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:202]: @tchughesiv Same here, display error msg to users?

File: src/components/instanceListPage.jsx [hunk 4]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+      if (namespace) {
+        promises.push(inventoriesFromRulesReview(namespace))
+      }
+    })
+    await Promise.all(promises)
+      .then((inventoryByNS) => {
+        inventoryByNS.forEach((inventoryArrays) => inventoryArrays.forEach((value) => inventoryItems.push(...value)))
+      })
+      .catch(function (error) {
+        console.error(error)
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:235]: @tchughesiv Same here, display error msg to users?

File: src/components/instanceListPage.jsx [hunk 5]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+      body: JSON.stringify(rulesBody),
+    }
+    listAllowed = await fetch('/api/kubernetes/apis/authorization.k8s.io/v1/selfsubjectaccessreviews', requestOpts)
+      .then(status)
+      .then(json)
+      .then((data) => {
+        return data.status.allowed
+      })
+      .catch(function (error) {
+        console.error(error)
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:271]: @tchughesiv Same here, display error msg to users?

File: src/components/instanceListPage.jsx [hunk 6]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+          .then((inventoryNames) => fetchInventories(inventoryNames, namespace))
+          .then((inventories) => {
+            return inventories
+          })
+      )
+    }
+    await Promise.all(promises)
+      .then((inventories) => (inventoryItems = inventories))
+      .catch(function (error) {
+        console.error(error)
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:331]: @tchughesiv Same here, display error msg to users?

File: src/components/instanceListPage.jsx [hunk 7]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+      })
+    return inventoryItems
+  }
+
+  const parseInventoryRulesReview = (responseJson) => {
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:336]: @tchughesiv This function looks very similar to the `parseTenantRulesReview ` function, maybe combine them?

File: src/components/instanceListPage.jsx [hunk 8]
```diff
@@ -131,20 +133,279 @@ const InstanceListPage = () => {
...
+      .then((data) => {
+        return data
+      })
+  }
+
+  function status(response) {
+    if (response.status >= 200 && response.status < 300) {
+      return Promise.resolve(response)
+    } else {
+      return Promise.reject(new Error(response.statusText))
```
Comment (yzhao583) [inline src/components/instanceListPage.jsx:403]: @tchughesiv Maybe show error msg to users? And ask them to try again?

**Review by yzhao583 — CHANGES_REQUESTED** [review]

File: src/components/instanceListPage.jsx (hunk 2 above)
Comment (tchughesiv) [reply src/components/instanceListPage.jsx:161]: I'm honestly not sure... is that the preference? to show the error to users?

Comment (tchughesiv) [reply src/components/instanceListPage.jsx]: done

**Review by yzhao583 — APPROVED** [review]

File: src/components/instanceListPage.jsx (hunk 8 above)
Comment (tchughesiv) [reply src/components/instanceListPage.jsx:386]: with the other changes i made, it seems this will be shown to users with try again. correct?

**Review by jeremyary — APPROVED** [review]

File: src/components/instanceListPage.jsx (hunk 8 above)
Comment (yzhao583) [reply src/components/instanceListPage.jsx:386]: Yes, I think so, but maybe we can do some tests.

**Comment by yzhao583** [comment]
@jeremyary @tchughesiv Is there any further change for this PR? Can we merge it?

**Pull request merged**